File: skill-test/projects/sample-cluster-java/src/main/java/cluster/java/cafe24/DummyCafe24Api.java
package cluster.java.cafe24;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public final class DummyCafe24Api implements AutoCloseable {

    public record ApiCallResult(
        int statusCode,
        String body,
        int bucketUsed,
        int bucketMax,
        int callRemainSeconds
    ) {
    }

    private final int bucketCapacity;
    private final int leakRatePerSecond;
    private final Map<String, AtomicInteger> bucketByMall = new ConcurrentHashMap<>();
    private final ScheduledExecutorService leakScheduler =
        Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "java-cafe24-leak");
            t.setDaemon(true);
            return t;
        });

    public DummyCafe24Api(int bucketCapacity, int leakRatePerSecond) {
        this.bucketCapacity = bucketCapacity;
        this.leakRatePerSecond = leakRatePerSecond;

        leakScheduler.scheduleAtFixedRate(() -> bucketByMall.values().forEach(level ->
                level.updateAndGet(current -> Math.max(0, current - leakRatePerSecond))),
            1, 1, TimeUnit.SECONDS);
    }

    public CompletionStage<ApiCallResult> call(String mallId, String word) {
        int level = bucketByMall.computeIfAbsent(mallId, k -> new AtomicInteger()).incrementAndGet();
        if (level > bucketCapacity) {
            bucketByMall.get(mallId).decrementAndGet();
            int over = level - bucketCapacity;
            int remain = (int) Math.ceil((double) over / leakRatePerSecond) + 1;
            return CompletableFuture.completedFuture(
                new ApiCallResult(429, "Too Many Requests", level, bucketCapacity, remain));
        }

        String body = "hello".equals(word) ? "world" : word;
        return CompletableFuture.completedFuture(
            new ApiCallResult(200, body, level, bucketCapacity, 0));
    }

    @Override
    public void close() {
        leakScheduler.shutdownNow();
    }
}

File: skill-test/projects/sample-cluster-dotnet/ClusterActors/DummyCafe24Api.cs
using System.Collections.Concurrent;

namespace ClusterActors;

public sealed class DummyCafe24Api : IDisposable
{
    public sealed record ApiCallResult(
        int StatusCode,
        string Body,
        int BucketUsed,
        int BucketMax,
        int CallRemainSeconds);

    private readonly int _bucketCapacity;
    private readonly int _leakRatePerSecond;
    private readonly ConcurrentDictionary<string, int> _bucketByMall = new();
    private readonly Timer _leakTimer;

    public DummyCafe24Api(int bucketCapacity, int leakRatePerSecond)
    {
        _bucketCapacity = bucketCapacity;
        _leakRatePerSecond = leakRatePerSecond;
        _leakTimer = new Timer(_ =>
        {
            foreach (var entry in _bucketByMall.ToArray())
            {
                var next = Math.Max(0, entry.Value - _leakRatePerSecond);
                _bucketByMall[entry.Key] = next;
            }
        }, null, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
    }

    public Task<ApiCallResult> CallAsync(string mallId, string word)
    {
        var level = _bucketByMall.AddOrUpdate(mallId, 1, (_, current) => current + 1);
        if (level > _bucketCapacity)
        {
            _bucketByMall.AddOrUpdate(mallId, 0, (_, current) => Math.Max(0, current - 1));
            var over = level - _bucketCapacity;
            var remain = (int)Math.Ceiling((double)over / _leakRatePerSecond) + 1;
            return Task.FromResult(new ApiCallResult(429, "Too Many Requests", level, _bucketCapacity, remain));
        }

        var body = word == "hello" ? "world" : word;
        return Task.FromResult(new ApiCallResult(200, body, level, _bucketCapacity, 0));
    }

    public void Dispose()
    {
        _leakTimer.Dispose();
    }
}
